app/models/query/results.rb

```diff
@@ -72,6 +83,31 @@ def order_option
 
   private
 
+  # For filtering on historic data, this returns the work packages
+  # matching the filters for any of the timestamps provided in the query.
+  #
+  def work_packages_matching_the_filters_at_any_of_the_given_timestamps
+    query.timestamps.collect do |timestamp|
+      WorkPackage.where(id: work_packages_relation.at_timestamp(timestamp))
+    end.reduce(:or)
+  end
+
+  # Returns an active-record relation that applies the filters to find the matching
+  # work packages.
+  #
+  # This can be chained with `.at_timestamp(...)` in order to search historic data
+  # as required for the baseline-comparison feature.
+  #
+  # https://community.openproject.org/projects/openproject/work_packages/26448
+  #
+  def work_packages_relation
+    work_package_scope
+      .where(query.statement)
+      .includes(all_includes)
```

Comment:
In the updated implementation, I'm still having `includes(all_includes)` in the `work_pacakge_scope`. Unused includes won't end up in the sql query.

```ruby
WorkPackage.includes(:project).to_sql
=> "SELECT \"work_packages\".* FROM \"work_packages\""

WorkPackage.includes(:project).references(:projects).to_sql
=> "SELECT \"work_packages\".\"id\" AS t0_r0, \"work_packages\".\"type_id\" AS t0_r1, \"work_packages\".\"project_id\" AS t0_r2, \"work_packages\".\"subject\" AS t0_r3, \"work_packages\".\"description\" AS t0_r4, \"work_packages\".\"due_date\" AS t0_r5, \"work_packages\".\"category_id\" AS t0_r6, \"work_packages\".\"status_id\" AS t0_r7, \"work_packages\".\"assigned_to_id\" AS t0_r8, \"work_packages\".\"priority_id\" AS t0_r9, \"work_packages\".\"version_id\" AS t0_r10, \"work_packages\".\"author_id\" AS t0_r11, \"work_packages\".\"lock_version\" AS t0_r12, \"work_packages\".\"done_ratio\" AS t0_r13, \"work_packages\".\"estimated_hours\" AS t0_r14, \"work_packages\".\"created_at\" AS t0_r15, \"work_packages\".\"updated_at\" AS t0_r16, \"work_packages\".\"start_date\" AS t0_r17, \"work_packages\".\"responsible_id\" AS t0_r18, \"work_packages\".\"budget_id\" AS t0_r19, \"work_packages\".\"position\" AS t0_r20, \"work_packages\".\"story_points\" AS t0_r21, \"work_packages\".\"remaining_hours\" AS t0_r22, \"work_packages\".\"derived_estimated_hours\" AS t0_r23, \"work_packages\".\"schedule_manually\" AS t0_r24, \"work_packages\".\"parent_id\" AS t0_r25, \"work_packages\".\"duration\" AS t0_r26, \"work_packages\".\"ignore_non_working_days\" AS t0_r27, \"projects\".\"id\" AS t1_r0, \"projects\".\"name\" AS t1_r1, \"projects\".\"description\" AS t1_r2, \"projects\".\"public\" AS t1_r3, \"projects\".\"parent_id\" AS t1_r4, \"projects\".\"created_at\" AS t1_r5, \"projects\".\"updated_at\" AS t1_r6, \"projects\".\"identifier\" AS t1_r7, \"projects\".\"lft\" AS t1_r8, \"projects\".\"rgt\" AS t1_r9, \"projects\".\"active\" AS t1_r10, \"projects\".\"templated\" AS t1_r11 FROM \"work_packages\" LEFT OUTER JOIN \"projects\" ON \"projects\".\"id\" = \"work_packages\".\"project_id\""
```

Having `includes(all_includes)` within the `work_package_scope` looks clean to me.

However, if you would prefer that, I would suggest that I split up the `include_columns` method so that in `filtered_work_packages` only the filter includes are added, and in `sorted_work_packages` only the sorting includes are added.